Scheduler class

dafyddstephenson opened the pull request:
## Summary
This PR:
- expands CStarSystem to add a `scheduler` property that determines the correct scheduler for the system and returns an appropriate instance of the correct `Scheduler` subclass (e.g. `SlurmScheduler`, `PBSScheduler`)
- Adds a `Queue` ABC (related to `Scheduler`)
- Adds a `scheduler_job` module with a `SchedulerJob` ABC and `SlurmJob` and ` PBSJob` subclasses. 

## More on `Scheduler`:
The `Scheduler` instance is instantiated with, e.g.:
```
SlurmScheduler(
                    queues=[SlurmQueue(name="regular", query_name="regular_1"),
                            SlurmQueue(name="shared"),
                            SlurmQueue(name="debug")],
                    primary_queue_name="regular",
                    queue_flag="qos",
                    other_scheduler_directives={"-C": "cpu"},
                )

```
Here:
- `queue_flag` refers to the name used to choose the queue in scheduler submissions, typically `partition` or `qos`
- `other_scheduler_directives` refers to system-specific choices that are always made by C-Star.

Properties and methods of note:
- `global_max_cpus_per_node`
- `global_max_mem_per_node_gb` 
both query the system to determine the absolute limits on cores per node and memory per node.

## More on `Queue`:
- `name` in `__init__` refers to the name used to submit a job to the queue, while `query_name` refers to the name used by the system when querying with the `sacctmgr` command. These default to being the same, but sometimes differ, as in the above example.

Properties of note (obtained by querying using `sacctmgr show qos` on slurm):
- `max_walltime`
- `max_nodes`
- `priority`

## More on `SchedulerJob`
A SchedulerJob instance is instantiated with, e.g.:
```
sj = SlurmJob(scheduler = cs.scheduler,
              commands = "echo 'hello world'",
              cpus = 1,
              account_key = os.environ.get("ACCOUNT_KEY"),
              script_path = Path.cwd()/"my_script.sh",
              run_path = Path.cwd(),
              job_name = "test_job",
              output_file = "test_file.out",
              queue_name = "shared",
              send_email = True,
              walltime = "00:10:00")
```
Properties and methods of note:
- `script` : dynamically creates a submission script from these parameters
- `submit` : submits the job to the scheduler and captures/returns the job id
- `id` : the job id mentioned above
- `status` : submitted, cancelled, running, complete
- `cancel` cancels the job after submitting to the scheduler

dafyddstephenson committed:
- define Scheduler, SlurmScheduler, Queue, SlurmQueue classes
- define basic properties on SlurmQueue and SlurmScheduler classes
- add basic logic to system.scheduler property

dafyddstephenson committed:
- define SchedulerJob and SlurmJob classes
- define basic properties on these classes
- add logic to submit slurm jobs and capture job IDs from sbatch output

dafyddstephenson committed:
test scheduler_job module, fix bugs, add 'cancel' method

dafyddstephenson committed:
add SlurmJob.status, SlurmJob.updates, SlurmScheduler.create_job

dafyddstephenson committed:
move Scheduler.create_job to scheduler_job.create

dafyddstephenson committed:
replace slurm-based logic in ROMSComponent.run(), update Case.run() and Component.run() to return Optional[SchedulerJob]

TomNicholas commented on cstar/base/scheduler.py:
```diff
@@ -0,0 +1,103 @@
+import subprocess
+
+if TYPE_CHECKING:
+    pass
```
```suggestion
```

TomNicholas commented on cstar/base/scheduler.py:
```diff
@@ -0,0 +1,103 @@
+        self.queue_names = [q.name for q in queues]
+        self.primary_queue_name = primary_queue_name
+        self.other_scheduler_directives = (
+            other_scheduler_directives if other_scheduler_directives is not None else {}
```
A docstring containing a complete example command would make the role of each of these arguments a lot clearer

TomNicholas commented on cstar/base/scheduler_job.py:
```diff
@@ -0,0 +1,287 @@
+## STIL TODO:
```
The prefix `scheduler_` in `scheduler_job.py` seems unnecessary? Or if this will only get used in conjunction with `scheduler.py`, perhaps they should both go in their own dedicated directory/module?

dafyddstephenson replied:
I have now restructured things so there's a dedicated `system` directory.
However `scheduler.py` and `environment.py` are sort of special cases in that they _define_ universal aspects of the system that should be available everywhere else in the package (via `CStarSystemManager`). As such, these are the only two modules that do not import from anywhere else in C-Star (to avoid circular importing). 

While seemingly closely related to `scheduler.py`, `scheduler_job.py` defines HPC job management-related methods, and should be able to import from elsewhere in C-Star (e.g. it requires access to system-level environment information, rather than providing this information). While on the surface `scheduler.py` and `scheduler_job.py` may seem related, this distinction means they should be kept separate.

The prefix is necessary to avoid ambiguity, as C-Star can run jobs without a scheduler.

TomNicholas replied:
That all makes sense, but then wouldn't a clearer name for `scheduler_job.py` be `scheduler/hpc.py`, and your standalone scheduler code that needs to be importable from anywhere can go in `scheduler/scheduler.py`?

> as C-Star can run jobs without a scheduler

Is it really a "job" if it's not using a scheduler? It's not a job in the normal sense of "HPC job submission queue". I would say C-Star can run a case without a scheduler, but not run the case as a job without a scheduler.

I mean this is not particularly important ofc - I don't want to waste your time.

dafyddstephenson replied:
I could do `cstar/system/scheduler.py` and `cstar/scheduler/job.py`? I think that would perhaps even make more sense.  May just seem a bit more fragmented, e.g. `job.py` would be the only module in `scheduler/`.

TomNicholas replied:
Happy with that also. We can also easily change this later as long as it's not too tangled.

dafyddstephenson replied:
I can do it now, no sweat

dafyddstephenson replied:
> I can do it now, no sweat

done in 5aef23d08166d293be4b81f23819dd11bd02755d and c05de1f3598a7a558e5d557c9fe69ddd8e6a1d72

TomNicholas commented on cstar/base/scheduler_job.py:
```diff
@@ -0,0 +1,287 @@
+class SchedulerJob(ABC):
+    @abstractmethod
+    def submit(self):
+        pass
```
Would an enumeration of the possible statuses of a job be useful here?

```suggestion
class JobStatus(Enum):
    PENDING = auto()
    RUNNING = auto()
    COMPLETED = auto()
    FAILED = auto()


class SchedulerJob(ABC):
    @abstractmethod
    def submit(self):
        pass
    
    @property
    def status(self) -> JobStatus:
        return self._status
```

dafyddstephenson replied:
Done in 907b767c24e687d7bf26448d6a668ffd674be5d4 and 23ab4d192f6299904fc71865b909f3311429ddd6

TomNicholas commented on cstar/base/scheduler_job.py:
```diff
@@ -0,0 +1,287 @@
+    @property
+    def status(self):
+        if self.id is None:
+            return "unsubmitted"
```
This really should be replaced with an `JobStatus` Enum to avoid "magic strings" (i.e. the string `"unsubmitted"` has a lot more meaning than the literal characters contained in it).

dafyddstephenson replied:
Yes, much better idea. This felt a bit wonky and I couldn't put my finger on why

dafyddstephenson replied:
Done in https://github.com/CWorthy-ocean/C-Star/commit/907b767c24e687d7bf26448d6a668ffd674be5d4 and https://github.com/CWorthy-ocean/C-Star/commit/23ab4d192f6299904fc71865b909f3311429ddd6

TomNicholas commented on cstar/base/system.py:
```diff
@@ -2,11 +2,36 @@
+            return self._scheduler
+
+        match self.name:
+            case "perlmutter":
```
Yeah we could be matching enums here.

dafyddstephenson replied:
Done in https://github.com/CWorthy-ocean/C-Star/commit/907b767c24e687d7bf26448d6a668ffd674be5d4 and https://github.com/CWorthy-ocean/C-Star/commit/23ab4d192f6299904fc71865b909f3311429ddd6

TomNicholas commented on cstar/base/system.py:
```diff
@@ -2,11 +2,36 @@
+    def scheduler(self) -> Scheduler:
+        """todo."""
+        if self._scheduler is not None:
+            return self._scheduler
```
I think this is a bit of an anti-pattern - shouldn't we always have a `Scheduler`, so the type of `_scheduler` shouldn't be `Optional`?

Could you instead have the private attribute always be set when the class is instantiated?

```python
class CStarSystem:
    _scheduler: Scheduler

    def __init__(self) -> None:
        self._scheduler = determine_scheduler_from_env()
```

dafyddstephenson replied:
see below

TomNicholas commented on cstar/case.py:
```diff
@@ -535,8 +535,8 @@ def run(
-        job_name="my_case_run",
-    ) -> None:
+        job_name=None,
+    ) -> Optional["SchedulerJob"]:
```
This type is confusing to me - does it return `None` if the job failed to submit? Or if it did submit but on a system where the job runs in the same process? This is kind of like returning a `dask.Delayed` object...

dafyddstephenson replied:
re. this and the above, comment, we do not always have a scheduler. Our integration tests on Github, for example, do not use a scheduler.

TomNicholas replied:
Okay. Just checking - does this mean that the syntax for running a case becomes different on different systems?

dafyddstephenson replied:
No, if `scheduler` is `None` then `Case.run()` will return `None` (and the execution will happen in the user's current python session). If there's a scheduler, it will return an appropriate `SchedulerJob` instance and the execution will happen somewhere else on the system.

TomNicholas replied:
Okay that's great, nice.

TomNicholas reviewed (commented):
This is cool. 

We should be trying really hard to keep the job scheduling and submission separate from the CStar-specific logic, because we maybe should be using one of the existing frameworks for job submission, e.g. https://parsl.readthedocs.io/en/stable/stubs/parsl.providers.SlurmProvider.html

---

I mentioned Enums a few times - I think now would be a good time to start using them. The `CStarEnvironment` class added in a previous PR could also potentially be an Enum.

dafyddstephenson commented:
Hi Tom, thanks for the feedback - will take a look tomorrow. Just a heads up that this is still WIP and there is significant work still to do - docstrings and such will come later (though I can provide ones for the things that are locked in place now if it will help read). Cheers!

dafyddstephenson committed:
remove residual scheduler-related attributes in environment

dafyddstephenson committed:
fix bug disallowing case of no scheduler

Commit:
update integration tests to use roms-tools v2.0

Commit:
update test_fixtures to reflect new input dataset list

Commit:
Merge branch 'main' into scheduler_class

Commit:
update unit tests

dafyddstephenson committed:
add PBSScheduler and PBSJob subclasses, methods, and properties

Commit:
fix typo (decorator order for abstract property)

dafyddstephenson committed:
remove residual PBS references in ROMSComponent

dafyddstephenson committed:
tidying up

Commit:
- Add unit tests for scheduler.py
- Fix bugs uncovered by unit tests
- Add str and repr methods for classes defined in scheduler.py

Commit:
begin adding unit tests to SchedulerJob

Commit:
create MockScheduler subclass for consistency instead of mock_scheduler fixture

Commit:
Begin adding unit tests to PBSJob

Commit:
add more PBSJob tests

Commit:
move repeat code into setup method for TestSchedulerJobBase

Commit:
fill out PBSJob unit tests and add test class for create_scheduler_job factory method

Commit:
complete unit tests for scheduler_job

Commit:
split CStarSystem unit tests into classes, add class for scheduler property

Commit:
Restructure:
- system.py,scheduler.py,scheduler_job.py,environment.py from cstar/base -> cstar/system
- rename CStarSystem class to CStarSystemManager
- rename cstar_system global to cstar_sysmgr
- rename system.py to manager.py (from cstar.system.manager import cstar_sysmgr, instead of from cstar.base.system import cstar_system)

Commit:
Merge branch 'main' into scheduler_class

Commit:
re-add unit tests for system missing from previous restructure commit

Commit:
separate scheduler_job unit tests into different modules for different subclasses

Commit:
change match blocks to use Enum in system/manager.py

Commit:
add JobStatus enum to system/scheduler_job.py

dafyddstephenson commented:
> maybe should be using one of the existing frameworks for job submission, e.g. https://parsl.readthedocs.io/en/stable/stubs/parsl.providers.SlurmProvider.html

I cannot tell you how hard I looked for a python package to take care of all this nonsense, how did you find this?! 
Right now I'm going to stick to what I've written here (it mostly just significantly improves what was already in), but we should definitely be looking into moving to Parsl later if it seems like it can do what we're looking for.

dafyddstephenson committed:
tweaks after initial Expanse testing

Commit:
change slurm export scheduler directive to ALL and remove in-script environment reconstruction

Commit:
update/simplify SLURM submission logic to restore compatibility with both Expanse and Perlmutter (via Jupyterhub)

Commit:
correct bug where base_model was looking in system environment variables rather than cstar.sysmgr.environment.environment_variables, failing to find MARBL_ROOT

Commit:
add 'requires_task_disribution' attribute to Scheduler class; update unit tests

Commit:
- REMOVE unused properties from Queue subclasses
- Split SlurmQueue into SlurmQOS and SlurmPartition
(in anticipation of eliminating 'queue_flag' logic and later adding QOS- and Partition-specific properties)

Commit:
remove queue_flag logic, now implicit in Queue subclass type

Commit:
cstar.system.scheduler_job -> cstar.scheduler.job

Commit:
move unit tests to reflect cstar.system.scheduler_job -> cstar.scheduler.job change

Commit:
- Add docstrings for scheduler.py
- Remove QueueFlag Enum
- change PBSQueue.max_walltime to a property and add _max_walltime private attr

Commit:
add docstrings for manager.py

Commit:
add docstrings for scheduler/job.py

Commit:
add docstrings for test_manager.py

Commit:
add docstrings for test_scheduler_job.py

Commit:
add remaining unit test docstrings

Commit:
correct decorator order for max_walltime property

Commit:
update unit tests to meet recent changes

Commit:
Merge branch 'main' into scheduler_class

dafyddstephenson committed:
update/add docpages, add 'documentation' attr to scheduler, update scheduler str fn

dafyddstephenson committed:
Merge branch 'main' into scheduler_class

Commit:
update unit tests

dafyddstephenson commented:
Took some extra time to sharpen this up more now it wasn't deadline-constrained, but got delayed as always by having to update the notebooks on Perlmutter. 

Ready to go now.

TomNicholas commented on cstar/case.py:
```diff
@@ -8,14 +8,14 @@
 
-# from cstar.base.environment import environment
-from cstar.base.system import cstar_system
+# from cstar.system.environment import environment
```
```suggestion
```

TomNicholas commented on cstar/case.py:
```diff
@@ -533,16 +533,22 @@ def pre_run(self) -> None:
-    ) -> None:
+        walltime=None,
+        queue=None,
+        job_name=None,
```
Type hints here would be nice - especially now as we are using some enums, it would be good to distinguish between types that are arbitrary strings and types that are one of a specific set of enum values.

dafyddstephenson replied:
done in fda70d91b2a12f8a42560b67a0d648542825d8ff

TomNicholas commented on cstar/case.py:
```diff
@@ -533,16 +533,22 @@ def pre_run(self) -> None:
         # TODO add more advanced logic for this
         # 20240807 - TN - set first component as main?
+
+        if (queue is None) and (cstar_sysmgr.scheduler is not None):
```
These brackets are redundant in python. I'm suprised the pre-commit linter hasn't removed them automatically. Another example of https://github.com/CWorthy-ocean/C-Star/issues/183

```suggestion
        if queue is None and cstar_sysmgr.scheduler is not None:
```

dafyddstephenson replied:
I think we've touched on this before - personal preference, if it's not a problem. Separating logical expressions into little packages like this makes it much more readable to me, especially as they get complex.

TomNicholas commented on cstar/case.py:
```diff
@@ -533,16 +533,22 @@ def pre_run(self) -> None:
+        if (queue is None) and (cstar_sysmgr.scheduler is not None):
+            queue = cstar_sysmgr.scheduler.primary_queue_name
+        if (walltime is None) and (cstar_sysmgr.scheduler is not None):
+            walltime = cstar_sysmgr.scheduler.get_queue(queue).max_walltime
```
So if `walltime` is `None` but scheduler is `None`, does `walltime` just stay as `None`? Is that okay?

This would be a case where having a public `walltime` variable and a private `_walltime` variable would allow mypy to catch possibly unwanted types sneaking through.

dafyddstephenson replied:
if `Scheduler` is `None` then `walltime` is never needed or used, I'd suggest `None` is the ideal way to signal this.

TomNicholas commented on cstar/case.py:
```diff
@@ -557,15 +563,17 @@ def run(
                     job_name=job_name,
                 )
+                return job_instance
+        return None
```
So if none of the components are ROMS this just returns `None`? Is that ever correct? Perhaps it should just error if it doesn't return before it gets to this line?

dafyddstephenson replied:
Fixed in [5565978](https://github.com/CWorthy-ocean/C-Star/pull/175/commits/5565978be084fa5a05ab14c6b3c4ca362c0fe729)

TomNicholas commented on cstar/roms/component.py:
```diff
@@ -880,6 +881,11 @@ def run(
+            queue = cstar_sysmgr.scheduler.primary_queue_name
+        if (walltime is None) and (cstar_sysmgr.scheduler is not None):
+            walltime = cstar_sysmgr.scheduler.get_queue(queue).max_walltime
+
```
Is this duplicated logic from the parent class?

TomNicholas commented on cstar/roms/component.py:
```diff
@@ -920,202 +926,111 @@ def run(
+                text=True,
+            )
+
+            # Process stdout line-by-line
```
All of this stdout processing should be abstracted into a function

dafyddstephenson replied:
Agree, this would be better as an issue though as this specific block only occurs in the absence of a scheduler. I was thinking of adding a parameter to the `SchedulerJob.updates` function that performs this same parsing rather than streaming raw output. Would be better in a separate PR though.

TomNicholas commented on cstar/roms/component.py:
```diff
@@ -920,202 +926,111 @@ def run(
+            # will exit with code 0 even if a fatal error occurs, see:
+            # https://github.com/CESR-lab/ucla-roms/issues/42
+
+            debugging = False  # Print raw output if true
```
This should either be removed or made into proper logging, used only when a global logging flag is set to debug level (xref https://github.com/CWorthy-ocean/C-Star/issues/88)

dafyddstephenson replied:
This is another example of cleaning that is needed (and is covered by 88) but is outside the scope of this PR (as `cstar_sysmgr.scheduler` is `None` for this block to be reached). There will be quite a few of these to get around to, this PR being the first and biggest example.

dafyddstephenson replied:
Revisiting this code this isn't actually covered by issue 88: the "debugging" flag here is used to prevent C-Star parsing ROMS stdout in case something has gone wrong with ROMS.
This can be handled more gracefully, though - #203 should cover it.

TomNicholas commented on cstar/roms/component.py:
```diff
@@ -920,202 +926,111 @@ def run(
+                    # Split the line by whitespace
+                    parts = line.split()
+
+                    # Check if there are exactly 9 parts and the first one is an integer
+                    if len(parts) == 9:
```
This is completely opaque without further comments giving context

TomNicholas replied:
It's also possible that whatever is being done here could be replaced with a regex match

dafyddstephenson replied:
again this relates to parsing the stdout of ROMS which is again a clean-up job independent of this PR.

dafyddstephenson replied:
Have opened #203 which should address this.

TomNicholas commented on cstar/scheduler/job.py:
```diff
@@ -0,0 +1,1001 @@
+            been submitted.
+        """
+
+        if self._id is None:
+            print("No Job ID found. Submit this job with SchedulerJob.submit()")
```
This property is nice, but this should probably raise (maybe a `RuntimeError`), because then the return type can always be `int`.

dafyddstephenson replied:
Could also set it as `0` or something as a placeholder if using a fixed type is particularly useful. I think a raise is a bit aggressive here, it's valid to query the id of a job that doesn't have one yet and I'd expect a gentle reminder rather than an error.

TomNicholas commented on cstar/scheduler/job.py:
```diff
@@ -0,0 +1,1001 @@
+
+        self.scheduler = scheduler
+        self.commands = commands
+        self.cpus = cpus
```
Currently the user could change these values by assigning to them after `SchedulerJob` is instantiated. Making them properties that returned private versions (e.g. `._scheduler`) would guard against this.

dafyddstephenson replied:
Done in 139092a86d34e8a4492693b88e9fe2c6dc298e18

TomNicholas reviewed (commented):
Splitting this review into parts as this PR is huge! I think the overall design seems good so the comments are mostly small.

dafyddstephenson commented:
> Splitting this review into parts as this PR is huge! I think the overall design seems good so the comments are mostly small.

Indeed - it is mostly docstrings, notebooks, and unit tests adding to the line count, most of the big-ticket changes are in 
`cstar/scheduler/job.py` and `cstar/system/scheduler.py` and should be quite digestible.

Generally speaking the thing that should be reviewed (as you've already begun previously) is the structure: if the structure checks out then the rest is largely just a reshuffling prioritising the separation of C-Star-related code from scheduler-related code. 

It will also be probably a lot easier to start by seeing how it's implemented in the doc pages (there is now a dedicated one related to this PR but I realise I didn't actually add it yet, notebook 5)

dafyddstephenson committed:
add notebook on using scheduler

dafyddstephenson committed:
update index.rst

dafyddstephenson commented:
[Notebook 5](https://c-star--175.org.readthedocs.build/en/175/5_handling_jobs_on_hpc_systems.html) now added.

I see a few comments related to [the parsing of the ROMS output ](https://github.com/CWorthy-ocean/C-Star/blob/cf35a94ba9d0d058765024e054e25eba695e67cb/cstar/roms/component.py#L979-L1031) so I'm not sure if this is showing up as new code in the diff - just as a heads up, this has been in for a good while, and isn't related to the scheduling refactor. #203 should cover this.

Commit:
add newlines before markdown lists in notebook 5

Commit:
change return None to raise in case Case.run does not find runnable component

Commit:
add typehints, rename queue -> queue_name in Case/Component.run

Commit:
rename queue -> queue_name in notebooks

Commit:
prevent certain SchedulerJob attributes from being set after init
